src/workerd/io/actor-cache-test.c++

```diff
@@ -1182,10 +1177,10 @@ KJ_TEST("ActorCache flush retry") {
     // last time, because it hasn't been further overwritten, and that delete from last time
     // wasn't actually committed.
     mockTxn->expectCall("delete", ws)
-        .withParams(CAPNP(keys = ["grault", "corge"]))
+        .withParams(CAPNP(keys = ["corge", "grault"]))
         .thenReturn(CAPNP(numDeleted = 2));
     mockTxn->expectCall("delete", ws)
-        .withParams(CAPNP(keys = ["quux", "baz"]))
```

Comment:
Hm, why aren't we including `quux` here anymore? The logic in the comment above still seems correct to me -- we need to ensure the delete for it is still sent because the previous attempt at deleting it never committed.

This looks like a bug, unless I'm misunderstanding.

Reply:
TODO for me -- review the rest of this commit more closely after this question has been sorted out

Reply:
This is a great call out, I think you're right.

Both before and after this change, the test is supposed to have the delete "succeed" (returning the count) and the transaction fail. We save the delete count [here](https://github.com/cloudflare/workerd/pull/1814/commits/1669f0d0e8c8beceeac378adfbc6f1d164a7504b#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2828), then because there's nothing left to do we mark the `CountedDelete` as having finished [right below](https://github.com/cloudflare/workerd/pull/1814/commits/1669f0d0e8c8beceeac378adfbc6f1d164a7504b#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2834).

Since the flush fails, we try again, but as we go through our `countedDeletes` we [check for those that have already finished](https://github.com/cloudflare/workerd/pull/1814/commits/1669f0d0e8c8beceeac378adfbc6f1d164a7504b#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2318-R2322) and skip them since they don't need to be deleted anymore. However, I suppose this violates the transaction assumptions, so instead we should probably be marking all the `CountedDelete`s as having been successfully deleted only after all the promises associated with the transaction [have completed](https://github.com/cloudflare/workerd/pull/1814/commits/1669f0d0e8c8beceeac378adfbc6f1d164a7504b#diff-52b552cc9d6477516b49e970aa23de236c7a1f7327b09c243c336445cbb2e18eR2895).

Will give it a try in the morning, I'm also curious to see if this would have consequences for `qux`, since we overwrite it with a `put` after the delete and I'm not totally sure what the expected behavior is. It looks like we shouldn't try to include it in the retried `CountedDelete` (judging by the test), but I suspect the current code doesn't handle that properly.

Reply:
https://github.com/cloudflare/workerd/pull/1814/commits/6d15b89f9ba8ec8395582196db46dd54a2260f44

I think this should address this particular issue, `quux` is now retried, though `qux` is not (since we overwrote it _and_ the first delete succeeded in giving us the count).

I've left a few additional TODOs for myself to think a bit more about, but I just want to get this off my local machine for now.